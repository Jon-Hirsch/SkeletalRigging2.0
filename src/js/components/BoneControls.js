import React from 'react';
import PropTypes from 'prop-types';
import dispatch from '../dispatch';

export default function BoneControls({ boneId, bones }) {
  const currentBone = bones.find(bone => bone.id === boneId);
  const children = bones.filter(bone => bone.parent === boneId);
  return (
    <div className="skeletal-rigging-bone-controls">
      <div className="bone-controls">
        <div>
          name:{' '}
          <input
            className="bone-name-input"
            type="text"
            value={currentBone.name}
            onChange={event =>
              updateBone(currentBone.id, 'name', event.target.value)}
          />
          x:{' '}
          <input
            className="bone-x-input"
            type="text"
            value={currentBone.x}
            onChange={event =>
              updateBone(currentBone.id, 'x', event.target.value)}
          />
          y:{' '}
          <input
            className="bone-y-input"
            type="text"
            value={currentBone.y}
            onChange={event =>
              updateBone(currentBone.id, 'y', event.target.value)}
          />
          length:{' '}
          <input
            className="bone-length-input"
            type="text"
            value={currentBone.boneLength}
            onChange={event =>
              updateBone(currentBone.id, 'boneLength', event.target.value)}
          />
          angle:{' '}
          <input
            type="text"
            className="bone-angle-input"
            value={currentBone.angle}
            onChange={event =>
              updateBone(currentBone.id, 'angle', event.target.value)}
          />
          <button
            className="add-bone-button"
            onClick={() => addBone(currentBone.id)}
          >
            Add Bone
          </button>
          {currentBone.parent && (
            <button
              className="remove-bone-button"
              onClick={() => removeBone(currentBone.id)}
            >
              Remove Bone
            </button>
          )}
        </div>
        <div />
      </div>
      {children.map(childBone => (
        <BoneControls key={childBone.id} boneId={childBone.id} bones={bones} />
      ))}
    </div>
  );
}

function addBone(id) {
  dispatch({
    type: 'ADD_BONE',
    parentId: id
  });
}

function removeBone(id) {
  dispatch({
    type: 'REMOVE_BONE',
    id
  });
}

function updateBone(id, field, value) {
  dispatch({
    type: 'UPDATE_BONE',
    id,
    field,
    value
  });
}

BoneControls.propTypes = {
  boneId: PropTypes.string.isRequired,
  bones: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      parent: PropTypes.string,
      x: PropTypes.string.isRequired,
      y: PropTypes.string.isRequired,
      boneLength: PropTypes.string.isRequired,
      angle: PropTypes.string.isRequired
    })
  ).isRequired
};
