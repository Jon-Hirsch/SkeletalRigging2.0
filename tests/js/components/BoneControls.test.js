import React from 'react';
import BoneControls from '../../../src/js/components/BoneControls';
import dispatch from '../../../src/js/dispatch';

jest.mock('../../../src/js/dispatch');

describe('BoneControls', () => {
  let shallowWrapper, props;

  function render() {
    return shallow(<BoneControls {...props} />);
  }

  beforeEach(() => {
    dispatch.mockClear();
    props = {
      boneId: 'body',
      bones: [
        {
          id: 'baseBone',
          name: 'baseBone',
          parent: null,
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        },
        {
          id: 'body',
          name: 'body',
          parent: 'baseBone',
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        },
        {
          id: 'child1',
          name: 'child1',
          parent: 'body',
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        },
        {
          id: 'child1',
          name: 'child1',
          parent: 'body',
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        },
        {
          id: 'child1',
          name: 'child1',
          parent: 'body',
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        }
      ]
    };
    shallowWrapper = render();
  });

  test('should match snapshot', () => {
    expect(shallowWrapper.getElement()).toMatchSnapshot();
  });

  test('it should render any child bones', () => {
    expect(shallowWrapper.find(BoneControls).length).toEqual(3);
  });

  test('it should not render a remove bone button if the bone is the base bone', () => {
    props.boneId = 'baseBone';
    shallowWrapper = render();
    const removeBoneButton = shallowWrapper
      .find('.bone-controls')
      .at(0)
      .find('.remove-bone-button');
    expect(removeBoneButton.length).toEqual(0);
  });

  test('clicking the add bone button should dispatch ADD_BONE', () => {
    shallowWrapper
      .find('.add-bone-button')
      .at(0)
      .simulate('click');
    expect(dispatch).toHaveBeenCalledWith({
      type: 'ADD_BONE',
      parentId: 'body'
    });
  });

  test('clicking the remove bone button should dispatch ADD_BONE', () => {
    shallowWrapper
      .find('.remove-bone-button')
      .at(0)
      .simulate('click');
    expect(dispatch).toHaveBeenCalledWith({
      type: 'REMOVE_BONE',
      id: 'body'
    });
  });

  test('changing the bone name should dispatch UPDATE_BONE', () => {
    shallowWrapper
      .find('.bone-name-input')
      .at(0)
      .simulate('change', { target: { value: 'newValue' } });
    expect(dispatch).toHaveBeenCalledWith({
      type: 'UPDATE_BONE',
      id: 'body',
      field: 'name',
      value: 'newValue'
    });
  });

  test('changing the bone x should dispatch UPDATE_BONE', () => {
    shallowWrapper
      .find('.bone-x-input')
      .at(0)
      .simulate('change', { target: { value: 'newValue' } });
    expect(dispatch).toHaveBeenCalledWith({
      type: 'UPDATE_BONE',
      id: 'body',
      field: 'x',
      value: 'newValue'
    });
  });

  test('changing the bone y should dispatch UPDATE_BONE', () => {
    shallowWrapper
      .find('.bone-y-input')
      .at(0)
      .simulate('change', { target: { value: 'newValue' } });
    expect(dispatch).toHaveBeenCalledWith({
      type: 'UPDATE_BONE',
      id: 'body',
      field: 'y',
      value: 'newValue'
    });
  });

  test('changing the bone length should dispatch UPDATE_BONE', () => {
    shallowWrapper
      .find('.bone-length-input')
      .at(0)
      .simulate('change', { target: { value: 'newValue' } });
    expect(dispatch).toHaveBeenCalledWith({
      type: 'UPDATE_BONE',
      id: 'body',
      field: 'boneLength',
      value: 'newValue'
    });
  });

  test('changing the bone angle should dispatch UPDATE_BONE', () => {
    shallowWrapper
      .find('.bone-angle-input')
      .at(0)
      .simulate('change', { target: { value: 'newValue' } });
    expect(dispatch).toHaveBeenCalledWith({
      type: 'UPDATE_BONE',
      id: 'body',
      field: 'angle',
      value: 'newValue'
    });
  });
});
