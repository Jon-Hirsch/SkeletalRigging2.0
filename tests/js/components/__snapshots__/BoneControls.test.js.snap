// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoneControls should match snapshot 1`] = `
<div
  className="skeletal-rigging-bone-controls"
>
  <div
    className="bone-controls"
  >
    <div>
      name:
       
      <input
        className="bone-name-input"
        onChange={[Function]}
        type="text"
        value="body"
      />
      x:
       
      <input
        className="bone-x-input"
        onChange={[Function]}
        type="text"
        value="300"
      />
      y:
       
      <input
        className="bone-y-input"
        onChange={[Function]}
        type="text"
        value="320"
      />
      length:
       
      <input
        className="bone-length-input"
        onChange={[Function]}
        type="text"
        value="200"
      />
      angle:
       
      <input
        className="bone-angle-input"
        onChange={[Function]}
        type="text"
        value="270"
      />
      <button
        className="add-bone-button"
        onClick={[Function]}
      >
        Add Bone
      </button>
      <button
        className="remove-bone-button"
        onClick={[Function]}
      >
        Remove Bone
      </button>
    </div>
    <div />
  </div>
  <BoneControls
    boneId="child1"
    bones={
      Array [
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "baseBone",
          "name": "baseBone",
          "parent": null,
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "body",
          "name": "body",
          "parent": "baseBone",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
      ]
    }
  />
  <BoneControls
    boneId="child1"
    bones={
      Array [
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "baseBone",
          "name": "baseBone",
          "parent": null,
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "body",
          "name": "body",
          "parent": "baseBone",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
      ]
    }
  />
  <BoneControls
    boneId="child1"
    bones={
      Array [
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "baseBone",
          "name": "baseBone",
          "parent": null,
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "body",
          "name": "body",
          "parent": "baseBone",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
        Object {
          "angle": "270",
          "boneLength": "200",
          "id": "child1",
          "name": "child1",
          "parent": "body",
          "x": "300",
          "y": "320",
        },
      ]
    }
  />
</div>
`;
